File: backend/app/schemas/short_term.py
"""短线机会雷达 API 模型。"""

from datetime import date
from typing import Literal

from pydantic import BaseModel, Field

ShortTermPeriod = Literal[
    "today", "current_week", "half_month", "current_month", "custom"
]


class MarketStrategyCardResponse(BaseModel):
    """指数与情绪组合策略卡。"""

    title: str = Field(description="卡片标题")
    index_strength: str = Field(description="指数强弱：strong/weak")
    emotion_strength: str = Field(description="情绪强弱：strong/weak")
    primary_strategy: str = Field(description="主策略")
    secondary_strategy: str = Field(description="辅助策略")
    operation_advice: str = Field(description="操作建议")
    focus_targets: list[str] = Field(default_factory=list, description="重点跟踪对象")
    rationale: list[str] = Field(default_factory=list, description="判断依据")
    formulas: list[str] = Field(
        default_factory=list, description="与判断依据一一对应的计算公式说明"
    )


class RefreshMeta(BaseModel):
    """刷新耗时与数据源信息。"""

    elapsed_ms: int = Field(default=0, description="刷新耗时（毫秒）")
    quote_source: str = Field(description="行情数据源")
    quote_attempts: list[str] = Field(default_factory=list, description="尝试过的数据源")
    quote_message: str = Field(default="", description="行情刷新说明")


class ShortTermOverviewResponse(BaseModel):
    """短线雷达概览。"""

    trade_date: date = Field(description="数据交易日")
    period: ShortTermPeriod = Field(default="today", description="统计周期")
    period_label: str = Field(default="当日", description="统计周期展示名")
    start_date: date = Field(description="统计开始日期")
    end_date: date = Field(description="统计结束日期")
    degraded: bool = Field(default=False, description="是否降级")
    missing_sources: list[str] = Field(default_factory=list, description="缺失数据源")
    market_emotion: str = Field(description="市场情绪")
    short_term_outlook: str = Field(description="短期展望")
    operation_advice: str = Field(description="操作建议")
    tracking_focus: list[str] = Field(default_factory=list, description="重点跟踪对象")
    core_conclusion: str = Field(description="核心结论")
    risk_signals: list[str] = Field(default_factory=list, description="风险信号")
    sector_count: int = Field(default=0, description="有效轮动板块数量")
    candidate_count: int = Field(default=0, description="候选数量")
    strategy_card: MarketStrategyCardResponse = Field(description="指数情绪策略卡")
    refresh_meta: RefreshMeta | None = Field(default=None, description="最近一次行情刷新元数据")


class FirstToSecondCandidateItem(BaseModel):
    """一进二候选股票。"""

    code: str = Field(description="股票代码")
    name: str = Field(description="股票名称")
    theme_name: str | None = Field(default=None, description="关联题材或行业")
    price: float | None = Field(default=None, description="最新价")
    market_cap: float | None = Field(default=None, description="总市值，单位亿元")
    float_market_cap: float | None = Field(default=None, description="流通市值，单位亿元")
    turnover_rate: float | None = Field(default=None, description="换手率")
    amount: float | None = Field(default=None, description="成交额，单位亿元")
    first_limit_up_at: str | None = Field(default=None, description="首次封板时间")
    open_board_count: int = Field(default=0, description="开板次数")
    score: int = Field(description="综合评分")
    decision: Literal["candidate", "watch", "excluded"] = Field(description="筛选结论")
    matched_rules: list[str] = Field(default_factory=list, description="命中规则")
    excluded_rules: list[str] = Field(default_factory=list, description="排除规则")
    risk_flags: list[str] = Field(default_factory=list, description="风险标签")
    catalysts: list[str] = Field(default_factory=list, description="催化说明")
    operation_advice: str = Field(description="操作建议")
    core_conclusion: str = Field(description="核心结论")


class FirstToSecondCandidateResponse(BaseModel):
    """一进二候选响应。"""

    trade_date: date = Field(description="目标交易日")
    previous_trade_date: date = Field(description="上一交易日")
    refreshed_at: str = Field(description="刷新时间")
    degraded: bool = Field(default=False, description="是否降级")
    missing_sources: list[str] = Field(default_factory=list, description="缺失数据源")
    candidates: list[FirstToSecondCandidateItem] = Field(
        default_factory=list, description="候选股票"
    )
    excluded_count: int = Field(default=0, description="排除数量")
    source_status: dict[str, str] = Field(default_factory=dict, description="数据源状态")


LifecycleStage = Literal[
    "germination", "fermentation", "climax", "divergence", "ebb"
]


class ShortTermSignalRefreshResponse(BaseModel):
    """短线信号全量刷新结果。"""

    trade_date: date
    status: Literal["success", "partial", "failed"]
    signal_count: int = 0
    dragon_tiger_count: int = 0
    sector_count: int = 0
    candidate_count: int = 0
    degraded: bool = False
    missing_sources: list[str] = Field(default_factory=list)
    source_status: dict[str, object] = Field(default_factory=dict)
    error_message: str | None = None


class SectorRotationItem(BaseModel):
    """板块轮动/主线题材项。"""

    theme_id: int
    theme_name: str
    board_kind: Literal["theme", "market", "indicator"] = Field(
        default="theme",
        description="theme=普通题材, market=市场表现, indicator=行情指标",
    )
    lifecycle_stage: LifecycleStage
    lifecycle_confidence: int = 0
    strength_score: int
    mainline_score: int
    risk_score: int = 0
    limit_up_count: int = 0
    failed_limit_up_count: int = 0
    summary: str = ""
    degraded: bool = False
    missing_metrics: list[str] = Field(default_factory=list)


class SectorRotationResponse(BaseModel):
    trade_date: date
    items: list[SectorRotationItem] = Field(default_factory=list)
    degraded: bool = False
    missing_sources: list[str] = Field(default_factory=list)


class LifecyclePoint(BaseModel):
    trade_date: date
    lifecycle_stage: LifecycleStage
    strength_score: int
    limit_quality_score: int | None = None
    flow_score: int | None = None
    leader_clarity_score: int | None = None
    breadth_score: int | None = None
    lifecycle_confidence: int = 0


class ThemeLifecycleResponse(BaseModel):
    theme_id: int
    days: int
    points: list[LifecyclePoint] = Field(default_factory=list)
    degraded: bool = False
    missing_sources: list[str] = Field(default_factory=list)

File: frontend/src/types/short-term.ts
export type MarketStrength = 'strong' | 'weak'
export type ShortTermPeriod = 'today' | 'current_week' | 'half_month' | 'current_month' | 'custom'
export type ShortTermPeriodStatusType = 'progress' | 'success' | 'error'
export type StrategyCardDataSource = 'database' | 'live'

export interface ShortTermPeriodStatus {
  type: ShortTermPeriodStatusType
  message: string
}

export interface MarketStrategyCardResponse {
  title: string
  index_strength: MarketStrength
  emotion_strength: MarketStrength
  primary_strategy: string
  secondary_strategy: string
  operation_advice: string
  focus_targets: string[]
  rationale: string[]
  /** 与 rationale 一一对应的计算公式说明 */
  formulas?: string[]
}

export interface RefreshMeta {
  elapsed_ms: number
  quote_source: string
  quote_attempts: string[]
  quote_message?: string
}

export interface ShortTermOverviewResponse {
  trade_date: string
  period: ShortTermPeriod
  period_label: string
  start_date: string
  end_date: string
  degraded: boolean
  missing_sources: string[]
  market_emotion: string
  short_term_outlook: string
  operation_advice: string
  tracking_focus: string[]
  core_conclusion: string
  risk_signals: string[]
  sector_count: number
  candidate_count: number
  strategy_card: MarketStrategyCardResponse
  refresh_meta?: RefreshMeta | null
}

export type FirstToSecondDecision = 'candidate' | 'watch' | 'excluded'

export interface FirstToSecondCandidateItem {
  code: string
  name: string
  theme_name: string | null
  price: number | null
  market_cap: number | null
  float_market_cap: number | null
  turnover_rate: number | null
  amount: number | null
  first_limit_up_at: string | null
  open_board_count: number
  score: number
  decision: FirstToSecondDecision
  matched_rules: string[]
  excluded_rules: string[]
  risk_flags: string[]
  catalysts: string[]
  operation_advice: string
  core_conclusion: string
}

export interface FirstToSecondCandidateResponse {
  trade_date: string
  previous_trade_date: string
  refreshed_at: string
  degraded: boolean
  missing_sources: string[]
  candidates: FirstToSecondCandidateItem[]
  excluded_count: number
  source_status: Record<string, string>
}

export type LifecycleStage =
  | 'germination'
  | 'fermentation'
  | 'climax'
  | 'divergence'
  | 'ebb'

export const LIFECYCLE_STAGE_LABEL: Record<LifecycleStage, string> = {
  germination: '萌芽',
  fermentation: '发酵',
  climax: '高潮',
  divergence: '分歧',
  ebb: '退潮',
}

export interface ShortTermSignalRefreshResponse {
  trade_date: string
  status: 'success' | 'partial' | 'failed'
  signal_count: number
  dragon_tiger_count: number
  sector_count: number
  candidate_count: number
  degraded: boolean
  missing_sources: string[]
  source_status: Record<string, unknown>
  error_message: string | null
}

export interface SectorRotationItem {
  theme_id: number
  theme_name: string
  /** theme=普通题材, market=市场表现, indicator=行情指标 */
  board_kind: 'theme' | 'market' | 'indicator'
  lifecycle_stage: LifecycleStage
  lifecycle_confidence: number
  strength_score: number
  mainline_score: number
  risk_score: number
  limit_up_count: number
  failed_limit_up_count: number
  summary: string
  degraded: boolean
  missing_metrics: string[]
}

export interface SectorRotationResponse {
  trade_date: string
  items: SectorRotationItem[]
  degraded: boolean
  missing_sources: string[]
}

export interface LifecyclePoint {
  trade_date: string
  lifecycle_stage: LifecycleStage
  strength_score: number
  limit_quality_score: number | null
  flow_score: number | null
  leader_clarity_score: number | null
  breadth_score: number | null
  lifecycle_confidence: number
}

export interface ThemeLifecycleResponse {
  theme_id: number
  days: number
  points: LifecyclePoint[]
  degraded: boolean
  missing_sources: string[]
}
